presto-rcfile/src/test/java/com/facebook/presto/rcfile/TestRcFileReaderManual.java

```diff
@@ -0,0 +1,333 @@
+                writeSegment(output, ImmutableList.of(ImmutableList.of(20, 22), ImmutableList.of(30, 33), ImmutableList.of(40, 44))),
+                writeSegment(output, ImmutableList.of(ImmutableList.of(100, 101, 102))));
+
+        assertFileSegments(output.slice(), segments);
+    }
+
+    @Test
+    public void testStartSync()
+            throws Exception
+    {
+        SliceOutput output = new DynamicSliceOutput(10 * 1024);
+
+        List<Segment> segments = ImmutableList.of(
+                writeSegment(output, ImmutableList.of()),
+                writeSegment(output, ImmutableList.of(ImmutableList.of(0, 2, 3, 4), ImmutableList.of(10, 12, 13))),
+                writeSegment(output, ImmutableList.of(ImmutableList.of(20, 22), ImmutableList.of(30, 33), ImmutableList.of(40, 44))),
+                writeSegment(output, ImmutableList.of(ImmutableList.of(100, 101, 102))));
+
+        assertFileSegments(output.slice(), segments);
+    }
+
+    private static void assertFileSegments(Slice file, List<Segment> segments)
+            throws IOException
+    {
+        // read whole file
+        List<Integer> allValues = segments.stream()
+                .map(Segment::getValues)
+                .flatMap(List::stream)
+                .collect(toList());
+        assertEquals(allValues, readValues(file, 0, file.length()));
+
+        for (int segmentIndex = 0; segmentIndex < segments.size(); segmentIndex++) {
+            Segment segment = segments.get(segmentIndex);
+
+            // whole segment
+            assertEquals(segment.getValues(), readValues(file, segment.getOffset(), segment.getLength()));
+            // first byte of segment
+            assertEquals(segment.getValues(), readValues(file, segment.getOffset(), 1));
+            // straddle segment start
+            assertEquals(segment.getValues(), readValues(file, segment.getOffset() - 1, 2));
+
+            // regions entirely within the the segment
+            assertEquals(ImmutableList.of(), readValues(file, segment.getOffset() + 1, 1));
+
+            for (Integer rowGroupOffset : segment.getRowGroupSegmentOffsets()) {
+                // segment header to row group start
+                assertEquals(segment.getValues(), readValues(file, segment.getOffset(), rowGroupOffset));
+                assertEquals(segment.getValues(), readValues(file, segment.getOffset(), rowGroupOffset - 1));
+                assertEquals(segment.getValues(), readValues(file, segment.getOffset(), rowGroupOffset + 1));
+
+                // regions entirely within the the row group
+                assertEquals(ImmutableList.of(), readValues(file, segment.getOffset() + rowGroupOffset, segment.getLength() - rowGroupOffset));
+            }
+
+            // combinations of previous segments
+            for (int previousSegmentIndex = 0; previousSegmentIndex < segmentIndex; previousSegmentIndex++) {
```

Comment:
I would pull this loop out into a separate 2D for loop. This is essentially testing all sub-ranges.